packages/core/src/node/core-process-manager.ts

```diff
@@ -0,0 +1,103 @@
+/********************************************************************************
+ * Copyright (C) 2021 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import * as cp from 'child_process';
+import { injectable } from 'inversify';
+
+@injectable()
+export class CoreProcessManager {
+
+    /**
+     * Synchronously send SIGTERM to a process and its children, recursively.
+     *
+     * @param ppid The process to kill along with its children (recursive).
+     */
+    terminateProcessTree(ppid: number): void {
+        if (process.platform === 'win32') {
+            this.win_taskkillTree(ppid);
+        } else {
+            for (const pid of this.unix_getChildrenRecursive(ppid)) {
```

Comment:
I very often use two libs and the following snippet to terminate a process (tree). It works for me reliably.
```ts
import * as psTree from 'ps-tree';
const kill = require('tree-kill');

psTree(daemonPid, function (_, children) {
  for (const { PID } of children) {
    kill(PID);
  }
  kill(daemonPid, () => process.exit());
});
```

Why do we need such customization? I guess someone in the Node.js world had the same issue, can simply use an existing lib?

Reply:
`psTree` is async and we need a sync process.

Reply:
Also when trying to spawn a script synchronously by doing `spawnSync(process.argv0, ...` it was crashing on Electron when packaged as an AppImage, hence why we are not spawning `psTree` in a subprocess.

Reply:
> it was crashing on Electron when packaged as an AppImage

What was the problem? Is it a confirmed issue?

I would recommend finding and using an existing solution for the process tree termination instead of reinventing the wheel. I am pretty sure we are not the first trying to achieve this functionality. How do others work around the problem?

Reply:
> What was the problem? Is it a confirmed issue?

The spawned Electron process kept crashing, or the script wasn't found. We couldn't pinpoint the actual reason, but my hypothesis was that the AppImage was unmounting itself and messed up the spawn process. It failed reliably.

> How do others work around the problem?

Well, as you mentioned `vscode-languageserver-node` spawns a bash script instead of a js one.

The only thing I prefer about the current solution is that I spawn only one process (`ps`) whereas the shell script spawns multiple processes (`pgrep`) in a recursive fashion.

Copying the .sh script might require a CQ though right?

Reply:
I'd rather keep it this way as we had confirmed that the current implementation works. We can retest to double check.

Reply:
I'd rather keep it this way as we had confirmed with @colin-grant-work that the current implementation works. We can retest to double check.